File: api/src/domain/ports/repositories/DatasetRepository.ts
import type { Dataset } from '../../entities/Dataset.js';
import type { DatasetId } from '../../value-objects/DatasetId.js';

/**
 * Port for Dataset persistence operations.
 * Infrastructure layer must implement this interface.
 */
export interface DatasetRepository {
  save(dataset: Dataset): Promise<void>;
  findById(id: DatasetId): Promise<Dataset | null>;
  findByUserId(userId: string): Promise<Dataset[]>;
  delete(id: DatasetId): Promise<void>;
  exists(id: DatasetId): Promise<boolean>;
}

File: worker/src/domain/ports/repositories/dataset_repository.py
"""Dataset repository port - interface for dataset persistence."""

from abc import ABC, abstractmethod
from uuid import UUID

from src.domain.entities.dataset import Dataset


class DatasetRepository(ABC):
    """Port for dataset persistence operations."""

    @abstractmethod
    async def find_by_id(self, id: UUID) -> Dataset | None:
        """Find a dataset by ID."""
        ...

    @abstractmethod
    async def save(self, dataset: Dataset) -> Dataset:
        """Save a dataset (create or update)."""
        ...

    @abstractmethod
    async def delete(self, id: UUID) -> bool:
        """Delete a dataset by ID."""
        ...

    @abstractmethod
    async def find_by_owner(
        self,
        owner_id: UUID,
        *,
        limit: int = 50,
        offset: int = 0,
    ) -> list[Dataset]:
        """Find datasets by owner."""
        ...
